File: src/main/java/main/commands/prefixcommands/audio/QueueCommand.java
package main.commands.prefixcommands.audio;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import main.audiohandlers.RobertifyAudioManager;
import main.commands.prefixcommands.CommandContext;
import main.commands.prefixcommands.ICommand;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.locale.LocaleManager;
import main.utils.locale.RobertifyLocaleMessage;
import main.utils.pagination.Pages;
import net.dv8tion.jda.annotations.ForRemoval;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.internal.utils.tuple.Pair;

import javax.script.ScriptException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Stack;
import java.util.concurrent.ConcurrentLinkedQueue;

@Deprecated @ForRemoval
public class QueueCommand implements ICommand {
    @Override
    public void handle(CommandContext ctx) throws ScriptException {
        final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(ctx.getGuild());
        final var queue = musicManager.getScheduler().queue;
        final Message msg = ctx.getMessage();

        GeneralUtils.setCustomEmbed(ctx.getGuild(), "Queue");

        if (queue.isEmpty()) {
            EmbedBuilder eb = RobertifyEmbedUtils.embedMessage(ctx.getGuild(), "There is nothing in the queue.");
            msg.replyEmbeds(eb.build()).queue();
            return;
        }

        sendQueue(ctx.getGuild(), queue, ctx.getChannel(), ctx.getAuthor());
        GeneralUtils.setDefaultEmbed(ctx.getGuild());
    }

    private void sendQueue(Guild guild, ConcurrentLinkedQueue<AudioTrack> queue, TextChannel channel, User user) {
        final List<AudioTrack> trackList = new ArrayList<>(queue);

        List<String> content = getContent(guild, queue, trackList);

        Pages.paginateMessage(channel, user, content, 10);
    }

    public List<String> getContent(Guild guild, ConcurrentLinkedQueue<AudioTrack> queue, List<AudioTrack> trackList) {
        List<String> content = new ArrayList<>();
        final var localeManager = LocaleManager.getLocaleManager(guild);
        for (int i = 0; i < queue.size(); i++) {
            final AudioTrack track = trackList.get(i);
            final AudioTrackInfo info = track.getInfo();
            content.add(localeManager.getMessage(RobertifyLocaleMessage.QueueMessages.QUEUE_ENTRY,
                    Pair.of("{id}", String.valueOf(i+1)),
                    Pair.of("{title}", info.title),
                    Pair.of("{author}", info.author),
                    Pair.of("{duration}", GeneralUtils.formatTime(track.getInfo().length))
            ));
        }
        return content;
    }

    public List<String> getContent(Guild guild, Stack<AudioTrack> pastTracks) {
        List<String> content = new ArrayList<>();
        final var trackList = new ArrayList<>(pastTracks.stream().toList());
        Collections.reverse(trackList);
        final var localeManager = LocaleManager.getLocaleManager(guild);
        for (int i = 0; i < pastTracks.size(); i++) {
            final AudioTrack track = trackList.get(i);
            final AudioTrackInfo info = track.getInfo();
            content.add(localeManager.getMessage(RobertifyLocaleMessage.QueueMessages.QUEUE_ENTRY,
                    Pair.of("{id}", String.valueOf(i+1)),
                    Pair.of("{title}", info.title),
                    Pair.of("{author}", info.author),
                    Pair.of("{duration}", GeneralUtils.formatTime(track.getInfo().length))
            ));
        }
        return content;
    }
    @Override
    public String getName() {
        return "queue";
    }

    @Override
    public String getHelp(String prefix) {
        return "Aliases: `"+GeneralUtils.listToString(getAliases())+"`\n" +
                "Shows all the queued songs";
    }

    @Override
    public List<String> getAliases() {
        return List.of("q");
    }
}

File: src/main/kotlin/main/commands/slashcommands/audio/QueueCommand.kt
package main.commands.slashcommands.audio

import main.audiohandlers.QueueHandler
import main.audiohandlers.RobertifyAudioManager
import main.audiohandlers.utils.author
import main.audiohandlers.utils.length
import main.audiohandlers.utils.title
import main.utils.GeneralUtils
import main.utils.RobertifyEmbedUtils.Companion.replyEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.locale.LocaleManager
import main.utils.locale.messages.GeneralMessages
import main.utils.locale.messages.QueueMessages
import main.utils.pagination.PaginationHandler
import main.utils.pagination.pages.queue.QueueItem
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent

class QueueCommand : AbstractSlashCommand(
    SlashCommand(
    name = "queue",
    description = "See all queued songs"
)
) {
    override suspend fun handle(event: SlashCommandInteractionEvent) {
        val guild = event.guild!!
        val musicManager = RobertifyAudioManager[guild]
        val queueHandler = musicManager.scheduler.queueHandler

        if (queueHandler.isEmpty) {
            event.replyEmbed {
                embed(GeneralMessages.NOTHING_IN_QUEUE)
            }.queue()
            return
        }

        PaginationHandler.paginateQueue(event)
    }

    suspend fun getContent(guild: Guild, queueHandler: QueueHandler): List<String> {
        val content = mutableListOf<String>()
        val trackList = queueHandler.contents.toMutableList()
        val localeManager = LocaleManager[guild]
        trackList.forEachIndexed { i, track ->
            content.add(localeManager.getMessage(QueueMessages.QUEUE_ENTRY,
                Pair("{id}", (i + 1).toString()),
                Pair("{title}", track.title),
                Pair("{author}", track.author),
                Pair("{duration}", GeneralUtils.formatTime(track.length))
            ))
        }
        return content
    }

    suspend fun getContent(guild: Guild, queueItems: List<QueueItem>): List<String> {
        val content = mutableListOf<String>()
        val localeManager = LocaleManager[guild]
        queueItems.forEach { track ->
            content.add(localeManager.getMessage(
                QueueMessages.QUEUE_ENTRY,
                Pair("{id}", track.trackIndex.toString()),
                Pair("{title}", track.trackTitle),
                Pair("{author}", track.artist),
                Pair("{duration}", GeneralUtils.formatTime(track.duration))
            ))
        }
        return content
    }

    suspend fun getPastContent(guild: Guild, queueHandler: QueueHandler): List<String> {
        val content = mutableListOf<String>()
        val trackList = queueHandler.previousTracksContents.toMutableList().reversed()
        val localeManager = LocaleManager[guild]
        trackList.forEachIndexed { i, track ->
            content.add(localeManager.getMessage(QueueMessages.QUEUE_ENTRY,
                Pair("{id}", (i + 1).toString()),
                Pair("{title}", track.title),
                Pair("{author}", track.author),
                Pair("{duration}", GeneralUtils.formatTime(track.length))
            ))
        }
        return content
    }

    override val help: String
        get() = "Shows all the queued songs"
}